lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,438 @@
+
+  ss << "MBlock: " << config.MBlock << ", NBlock: " << config.NBlock
+     << ", KBlock: " << config.KBlock << ", MThreads: " << config.MThreads
+     << ", NThreads: " << config.NThreads << ", KThreads: " << config.KThreads
+     << ", innerMostMBlock: " << config.innerMostMBlock
+     << ", innerMostNBlock: " << config.innerMostNBlock
+     << ", innerMostKBlock: " << config.innerMostKBlock;
+  return ss;
+}
+
+template <typename T>
+static llvm::raw_ostream &operator<<(llvm::raw_ostream &ss,
+                                     std::vector<T> arry) {
+  ss << "[";
+  for (auto [idx, a] : llvm::enumerate(arry)) {
+    if (idx != 0) {
+      ss << ", ";
+    }
+    ss << a;
+  }
+  ss << "]";
+  return ss;
+}
+
+// generate the candidate for the block size(factor of `num`, pow of 2 which is
+// less than `num`)
+std::vector<uint32_t>
+getCandidate(uint32_t num, uint32_t floor,
+             uint32_t ceil = std::numeric_limits<uint32_t>::max()) {
+  // factor
+  std::vector<uint32_t> candidates;
+  for (uint32_t i = 1; i <= num; i++) {
+    if (num % i == 0 && i <= ceil && i >= floor) {
+      candidates.push_back(i);
+    }
+  }
+  // the pow of 2
+  uint32_t candidate = 1U;
+  while (candidate < num && candidate <= ceil && candidate >= floor) {
+    candidates.push_back(candidate);
+    candidate *= 2;
+  }
+  // remove duplicate candidates
+  std::sort(candidates.begin(), candidates.end());
+  candidates.erase(std::unique(candidates.begin(), candidates.end()),
+                   candidates.end());
+  return candidates;
+}
+
+// check if the config is valid
+bool isValidConfig(const MatmulConfig &config, SystemDesc &sysDesc,
+                   ArrayRef<uint32_t> shape) {
+  if (config.innerMostMBlock == 0 || config.innerMostNBlock == 0 ||
+      config.innerMostKBlock == 0) {
+    return false;
+  }
+
+  if (shape[0] % config.innerMostMBlock != 0 ||
+      shape[1] % config.innerMostNBlock != 0 ||
+      shape[2] % config.innerMostKBlock != 0) {
+    return false;
+  }
+
+  return true;
+}
+
+// check if the threads are valid
+bool validateThreads(ArrayRef<uint32_t> threads, SystemDesc &sysDesc) {
+  uint32_t numThreads = sysDesc.getNumThreads();
+  uint32_t actualThreads = 1U;
+  for (uint32_t t : threads) {
+    actualThreads *= t;
+  }
+  return actualThreads == numThreads;
+}
+
+// calculate the cost of the hardware efficiency(whether the vector register is
+// fully utilized)
+double hardwareEfficiencyCost(linalg::LinalgOp &linalgOp,
+                              ArrayRef<uint32_t> shape,
+                              const MatmulConfig &config, SystemDesc &sysDesc) {
+  llvm::TypeSize dtypeSize = DataLayout().getTypeSizeInBits(
+      ShapeAdaptor(linalgOp.getDpsInputs()[1].getType()).getElementType());
+  size_t maxVectorLength = sysDesc.getMaxVectorLength() / dtypeSize;
+  double cost = (maxVectorLength - config.innerMostMBlock % maxVectorLength) %
+                    maxVectorLength * 1.0 / config.innerMostMBlock +
```

Comment:
Not sure if this is the ideal algo to calculate the register utilization, as the computation of matmul is not always loading full register width from M or K dimension (or it might depends on impl), while most likely it's depending on `N` for the dotproduct? Do you think this cost factor really matters?

Reply:
The cost model is still under-tuning and this could be further adjusted. This cost is used to let the inner block be as close to the register length or its multiple as possible. It is hard to know how brgemm utilizes register at the linalg level so here just an approximation. And in the current benchmark, this is good for brgemm performance at most cases.

Reply:
I'm considering the AMX register, for BF16 datatype, this cost function prefers multiple of (32,32,32), while multiple of (64,64,64) for INT8 datatype, I assume 32,64,32 would also an option for INT8. Let's revise this with more benchmark result.

Reply:
Yes, a smaller NBlock and MBlock are also enough for amx case. I add a todo here to further adjust this cost model.